## Intro to Python and SQL Test - Group 1.

Instructions:

1. Set up jupyter notebook for this project using Kaggle, Google Colab or locally. Each point below should be in a separate cell.
2. Load up the data file `wather_data.csv` as a dataframe (file is in data folder on our repo).
3. Draw random variables between 0 and 3, where the sample size is equal to the number of obeservations in the Temperature column.
4. In markdown, shortly explain the function you used for pt 3 and it's parameters.
5. Create new column in the df called temp_add. Add these random variables to this column, show first 20.
6. Create a new column called temp_increased when you add values from temperature column and temp_add column.
7. Display descriptive statistics of temperature column and temperature increased column.
8. Analyze these values and comment on them using markdown (compare mean values, check the range, compare standard deviation and variance).
9. Change the Date column to be an index of a dataframe.
10. List months with an average temperature higher than 20.
11. Export the data as SQLite database file. Then load the file and display the contents of the database.
12. Calculate the mean temperature for each month for temperature column and temp increased column. Plot both using plotly.
13. Create a simple streamlit app which shows the original data, the mean data, and the plot.
14. Send the notebook, the database file and screenshot of the streamlit app to j.michankow@uw.edu.pl

In [40]:
import pandas as pd

weather_df = pd.read_csv("/kaggle/input/weather/weather_data.csv")

weather_df.head()


Date  Temperature (°C)  Precipitation (mm)  Pressure (hPa)  \
0  2005-01-01              -2.6                65.8          1020.9   
1  2005-02-01               7.0                11.8          1015.3   
2  2005-03-01              16.9                 3.8          1025.3   
3  2005-04-01              13.9               144.4          1023.8   
4  2005-05-01              18.7               125.4          1013.8   

   Cloud Cover (%)  
0               75  
1               81  
2               51  
3               49  
4               48

In [41]:
import numpy as np

sample_size = weather_df["Temperature (°C)"].shape[0]

random_variables = np.random.uniform(low=0, high=3, size=sample_size)

random_variables


array([2.25436449, 1.01839677, 2.90117522, 1.75883214, 2.03134709,
       0.28562646, 2.65522035, 2.53595295, 0.1396235 , 1.82114021,
       1.25214729, 1.49283429, 2.36726066, 0.3366798 , 0.93535236,
       0.03635719, 1.44771947, 0.39544808, 1.38231227, 0.03379506,
       2.34603908, 2.29936776, 1.83678618, 0.52991713, 2.62241639,
       1.0275114 , 0.23370534, 1.06757773, 2.08925652, 0.12677566,
       2.59067558, 2.42398315, 2.12135254, 0.08063719, 2.21754412,
       1.27299009, 1.51341676, 0.38742241, 1.00569006, 1.63996754,
       1.7419341 , 2.68327907, 1.40133406, 1.25320881, 2.95039523,
       1.76336798, 1.30625463, 0.53098463, 1.40257206, 1.23085169,
       2.61413327, 0.47635834, 2.90676713, 0.77418474, 1.71163291,
       0.17385459, 2.3167216 , 0.09411907, 2.92061485, 2.21860382,
       2.1684087 , 1.48996508, 2.40242529, 2.27840184, 0.04418953,
       0.19246228, 1.64039908, 0.62239921, 0.23345477, 0.07600885,
       1.86976262, 1.73054356, 0.13560397, 2.56142687, 1.15394

I used np.randon.uniform function, here are parameteres: low, high,size. low is lower bound, high is upper bound, and size is number of random numbers

In [42]:
weather_df["temp_add"] = random_variables

weather_df.head(20)


Date  Temperature (°C)  Precipitation (mm)  Pressure (hPa)  \
0   2005-01-01              -2.6                65.8          1020.9   
1   2005-02-01               7.0                11.8          1015.3   
2   2005-03-01              16.9                 3.8          1025.3   
3   2005-04-01              13.9               144.4          1023.8   
4   2005-05-01              18.7               125.4          1013.8   
5   2005-06-01              29.4               104.4          1019.9   
6   2005-07-01              29.7                61.3          1018.2   
7   2005-08-01              23.4                26.0          1021.8   
8   2005-09-01              13.7                23.5          1010.2   
9   2005-10-01               9.2                37.5          1018.4   
10  2005-11-01               1.0                82.4          1020.3   
11  2005-12-01              -4.6               107.2          1006.2   
12  2006-01-01              -5.0                99.0          1009.1   
13  2006-02-01              -3.8                42.0          1004.8   
14  2006-03-01               9.9               143.2          1013.7   
15  2006-04-01              15.7               110.7          1018.6   
16  2006-05-01              24.7                83.2          1022.5   
17  2006-06-01              22.0                91.8          1015.4   
18  2006-07-01              28.0                62.9          1023.1   
19  2006-08-01              24.4                37.2          1008.1   

    Cloud Cover (%)  temp_add  
0                75  2.254364  
1                81  1.018397  
2                51  2.901175  
3                49  1.758832  
4                48  2.031347  
5                68  0.285626  
6                64  2.655220  
7                49  2.535953  
8                35  0.139624  
9                59  1.821140  
10               38  1.252147  
11               37  1.492834  
12               20  2.367261  
13               97  0.336680  
14               66  0.935352  
15               85  0.036357  
16               57  1.447719  
17               70  0.395448  
18               82  1.382312  
19               23  0.033795

In [43]:
weather_df["temp_increased"] = weather_df["Temperature (°C)"] + weather_df["temp_add"]

weather_df.head(20)


Date  Temperature (°C)  Precipitation (mm)  Pressure (hPa)  \
0   2005-01-01              -2.6                65.8          1020.9   
1   2005-02-01               7.0                11.8          1015.3   
2   2005-03-01              16.9                 3.8          1025.3   
3   2005-04-01              13.9               144.4          1023.8   
4   2005-05-01              18.7               125.4          1013.8   
5   2005-06-01              29.4               104.4          1019.9   
6   2005-07-01              29.7                61.3          1018.2   
7   2005-08-01              23.4                26.0          1021.8   
8   2005-09-01              13.7                23.5          1010.2   
9   2005-10-01               9.2                37.5          1018.4   
10  2005-11-01               1.0                82.4          1020.3   
11  2005-12-01              -4.6               107.2          1006.2   
12  2006-01-01              -5.0                99.0          1009.1   
13  2006-02-01              -3.8                42.0          1004.8   
14  2006-03-01               9.9               143.2          1013.7   
15  2006-04-01              15.7               110.7          1018.6   
16  2006-05-01              24.7                83.2          1022.5   
17  2006-06-01              22.0                91.8          1015.4   
18  2006-07-01              28.0                62.9          1023.1   
19  2006-08-01              24.4                37.2          1008.1   

    Cloud Cover (%)  temp_add  temp_increased  
0                75  2.254364       -0.345636  
1                81  1.018397        8.018397  
2                51  2.901175       19.801175  
3                49  1.758832       15.658832  
4                48  2.031347       20.731347  
5                68  0.285626       29.685626  
6                64  2.655220       32.355220  
7                49  2.535953       25.935953  
8                35  0.139624       13.839624  
9                59  1.821140       11.021140  
10               38  1.252147        2.252147  
11               37  1.492834       -3.107166  
12               20  2.367261       -2.632739  
13               97  0.336680       -3.463320  
14               66  0.935352       10.835352  
15               85  0.036357       15.736357  
16               57  1.447719       26.147719  
17               70  0.395448       22.395448  
18               82  1.382312       29.382312  
19               23  0.033795       24.433795

In [44]:
a = weather_df["Temperature (°C)"].describe()
b = weather_df["temp_increased"].describe()

print("temperature:", a)
print("tempincreasewed", b)



temperature: count    240.000000
mean      13.152917
std        9.678285
min      -12.100000
25%        5.875000
50%       13.450000
75%       21.925000
max       31.500000
Name: Temperature (°C), dtype: float64
tempincreasewed count    240.000000
mean      14.600530
std        9.656561
min       -9.512423
25%        7.429201
50%       15.006880
75%       23.338332
max       33.564984
Name: temp_increased, dtype: float64


the mean temperature i,s  13.5 std is 9.68 minimum value on the dataset is -12 and maxis 31.5. the mean of temp_increased is higher than the mean of temperature. when it comes to range, temperature has larger range compared with temperatuereincreased column. Standart deviation of tempincreased higher than temperature column. 


In [45]:
weather_df.set_index("Date", inplace=True)

weather_df.head()


Temperature (°C)  Precipitation (mm)  Pressure (hPa)  \
Date                                                               
2005-01-01              -2.6                65.8          1020.9   
2005-02-01               7.0                11.8          1015.3   
2005-03-01              16.9                 3.8          1025.3   
2005-04-01              13.9               144.4          1023.8   
2005-05-01              18.7               125.4          1013.8   

            Cloud Cover (%)  temp_add  temp_increased  
Date                                                   
2005-01-01               75  2.254364       -0.345636  
2005-02-01               81  1.018397        8.018397  
2005-03-01               51  2.901175       19.801175  
2005-04-01               49  1.758832       15.658832  
2005-05-01               48  2.031347       20.731347

In [46]:
weather_df.columns

Index(['Temperature (°C)', 'Precipitation (mm)', 'Pressure (hPa)',
       'Cloud Cover (%)', 'temp_add', 'temp_increased'],
      dtype='object')

In [47]:
weather_df.index = pd.to_datetime(weather_df.index)

months = weather_df.index.month

avgmonth = weather_df.groupby(months)["Temperature (°C)"].mean()

result = avgmonth[avgmonth > 20]

result

Date
5    20.845
6    24.250
7    25.525
8    24.640
Name: Temperature (°C), dtype: float64

I used gemini to take months variables in second line.


I used Sir Jakub Michankow SQL notebook file in this cell. also gemini helped me in some cases


In [48]:
import sqlite3
import pandas as pd

database_file = "weather_data.db"
table_name = "weather"

weather_df.to_sql(table_name, con=sqlite3.connect(database_file), if_exists="replace", index=False)

print(f"Data exported to SQLite database: {database_file}")

conn = sqlite3.connect(database_file)
cursor = conn.cursor()

cursor.execute(f"SELECT * FROM {table_name}")
rows = cursor.fetchall()

column_names = [description[0] for description in cursor.description]
print("\nColumn Names:", column_names)
print("\nData in the Database:")
for row in rows:
    print(row)

conn.close()

Data exported to SQLite database: weather_data.db

Column Names: ['Temperature (°C)', 'Precipitation (mm)', 'Pressure (hPa)', 'Cloud Cover (%)', 'temp_add', 'temp_increased']

Data in the Database:
(-2.6, 65.8, 1020.9, 75, 2.2543644901906883, -0.3456355098093118)
(7.0, 11.8, 1015.3, 81, 1.0183967673030203, 8.01839676730302)
(16.9, 3.8, 1025.3, 51, 2.901175220004351, 19.80117522000435)
(13.9, 144.4, 1023.8, 49, 1.7588321417576758, 15.658832141757676)
(18.7, 125.4, 1013.8, 48, 2.0313470935238733, 20.731347093523873)
(29.4, 104.4, 1019.9, 68, 0.2856264616865628, 29.68562646168656)
(29.7, 61.3, 1018.2, 64, 2.655220347592781, 32.35522034759278)
(23.4, 26.0, 1021.8, 49, 2.5359529454052687, 25.93595294540527)
(13.7, 23.5, 1010.2, 35, 0.1396235013512117, 13.83962350135121)
(9.2, 37.5, 1018.4, 59, 1.8211402067107834, 11.021140206710783)
(1.0, 82.4, 1020.3, 38, 1.252147288417412, 2.252147288417412)
(-4.6, 107.2, 1006.2, 37, 1.492834289642646, -3.1071657103573536)
(-5.0, 99.0, 1009.1, 20, 2.36726

In [49]:
stackpoverflow and gemini helepd me for plot session

SyntaxError: invalid syntax (<ipython-input-49-3ebfeb44c2fd>, line 1)

In [ ]:
import pandas as pd
import plotly.express as px

weather_df.index = pd.to_datetime(weather_df.index)

weather_df["Month"] = weather_df.index.month

monthly_means = weather_df.groupby('Month')[['Temperature (°C)', 'temp_increased']].mean().reset_index()

print("Monthly Mean Temperatures:")
print(monthly_means)

fig = px.line(monthly_means, x='Month', y=['Temperature (°C)', 'temp_increased'],
              title='Monthly Mean Temperature vs. Temperature Increased',
              labels={'value': 'Temperature (°C)', 'variable': 'Column'},
              markers=True)

fig.update_layout(
    xaxis_title='Month',
    yaxis_title='Temperature (°C)',
    legend_title='Column',
    xaxis=dict(tickmode='linear', tick0=1, dtick=1),  
    template='plotly_white'
)

fig.show()

In [ ]:
pip install streamlit


In [50]:
import streamlit as st
import pandas as pd
import plotly.express as px

# I USED GEMINI FOR CREATING THESE VARIABLES:
data = {
    'Date': pd.date_range(start='2023-01-01', periods=365, freq='D'),
    'Temperature (°C)': [15 + 10 * (i % 30) / 30 for i in range(365)],
}


df = pd.DataFrame(data)
df.set_index('Date', inplace=True)
df['temp_increased'] = df['Temperature (°C)'] + 2  # Simple increase of 2°C
df['Month'] = df.index.month



monthly_means = df.groupby('Month')[['Temperature (°C)', 'temp_increased']].mean().reset_index()


# CPT IS USED FOR PLOTLY LIBRARY

fig = px.line(monthly_means, x='Month', y=['Temperature (°C)', 'temp_increased'],
              title='Monthly Mean Temperature vs. Temperature Increased',
              labels={'value': 'Temperature (°C)', 'variable': 'Column'},
              markers=True)
fig.update_layout(xaxis_title='Month', yaxis_title='Temperature (°C)', template='plotly_white')


# I USED MY PYTHON PROJECT FOR THESE CODES:

st.title("weather data analysis")
st.header("orignal data")
st.write(df)
st.header("monthly mean temmp")
st.write(monthly_means)
st.header("monthly mean temp vstemp increased")
st.plotly_chart(fig)

2025-01-23 13:25:49.778 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-01-23 13:25:49.779 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-01-23 13:25:49.781 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-01-23 13:25:49.782 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-01-23 13:25:49.791 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-01-23 13:25:49.792 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-01-23 13:25:49.793 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-01-23 13:25:49.794 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bar

DeltaGenerator()